### ISSUE TITLE
Using multiple ThemeProvider with components prop at different levels can break the global theme of a website

### ISSUE THREAD

**Issue by soumikcv** [siddharthkp/react-ui#44]
I found a bug in the [react-ui website](https://react-ui.dev/core-concepts/constraints-based-design) where this page broke on switching to dark mode. It also happens in the [customizing-components page](https://react-ui.dev/core-concepts/customising-components).
![image](https://user-images.githubusercontent.com/18064965/79515812-496ba680-8067-11ea-9358-42663dba6753.png)

This is happening because of the ThemeProvider being used in pages/core-concepts/constraints.js with light tokens and light components. When the components prop is passed, the Global styles from the theme gets injected because of the way ThemeProvider is implemented.

```
function ThemeProvider({
  tokens = base.tokens,
  components = base.components,
  theme = {}, // as a combined fallback, not recommended but you know
  ...props
}) {

   // ...implementation code

  return (
    <Provider theme={combinedTheme} {...props}>
      <Global styles={interpolate(components.Global, combinedTheme)} />
      {props.children}
    </Provider>
  )
}
```

I assumed the motivation behind using the ThemeProvider in this page is to keep the Examples in light theme while the rest stay in dark(although idk why it's at the top level of this page). 

(There are other pages using ThemeProvider without passing the component prop, hence they do not face this issue)

This however breaks the website by injecting the light global theme. 

### My solutions

1. A simple way would be to exclude `Global` key from component theme before passing it as components prop

2. It might be a common use-case to force some components to use certain theme in which case we can just conditionally inject global style based on a prop say `injectGlobalStyle : boolean` or some kind of a reusable component which does the same.

**Comment by siddharthkp** [comment]
> I assumed the motivation behind using the ThemeProvider in this page is to keep the Examples in light theme while the rest stay in dark 

Yep, that's the motivation

I see two problems here :)

1. react-ui docs: 

	> although idk why it's at the top level of this page
   
    Yep, that's the problem here 👍 We should wrap only the examples, not the entire page

2. Nested `ThemeProvider`:

   When you're using multiple theme providers on a page, the intention is to create a section of the page that follows the theme.
   
   In the case of a nested themeprovider, you still want the global styles to apply, but you want them on the `root` of the section so that it cascades down. 

   We could replace `body` (root of the page) to `root` in [the theme specification - that sets the expectation that it gets applied to an element at the themeprovider root. We probably need to create an additional div to make this work.

   ```diff
   Global: {
   -  body: {
       background: 'white',
       color: 'text.body'
     }
   }
   ```

   It's fair to expect styles outside the root are applied to

**Comment by soumikcv** [comment]
Your approach makes a lot more sense. Now is a bad time in the world to hide symptoms 🙃

The change in Global API in this format looks like a good idea as it'll also enable a more elegant(?) implementation of the theme-provider (with no interpolation here yay!)

```diff
Global: {
  background: 'white',
  color: 'text.body',
  a: {
    color: 'text.link'
  }
}
```
```diff
function ThemeProvider({
  tokens = base.tokens,
  components = base.components,
  theme = {}, // as a combined fallback, not recommended but you know
  ...props
}) {

   // ...implementation code

  return (
    <Provider theme={combinedTheme} {...props}>
-      <Global styles={interpolate(components.Global, combinedTheme)} />
+      <Element as="div" component="Global">
+       {props.children}
+      </Element>
      {props.children}
    </Provider>
  )
```

> Although I can imagine the preference to keep it explicit

If you want to keep it explicit, maybe changing the name from Global to Root could make sense because it's not really Global anymore.

```diff
-Global: {
+Root: {
  background: 'white',
  color: 'text.body',
  a: {
    color: 'text.link'
  }
}
```

I'm assuming there still needs to be a Global style tag for the ` { ':root': { '--reach-menu-button': 1 } }` although idk what that does yet. 

However, adding multiple ThemeProviders adds multiple 
`<style data-emotion="css-global">:root{--reach-menu-button:1;}</style>`


Another smol implementation detail in this case would be the root div should have `overflow: auto` to avoid this 
![image](https://user-images.githubusercontent.com/18064965/79559851-13104480-80c4-11ea-84a0-6f44326cb6c9.png)

**Comment by siddharthkp** [comment]
Good catch!

We would need to support both root and global styles.

I wonder if we can be cheeky about it and identify if a certain style should be on theme provider root or document root.

`body, document, :root` are tags that are meant for the whole page and should be attached all the way up, but the rest of the things don't need to.

**Comment by soumikcv** [comment]
Aight! Makes sense. 

I wanted to take this one up but it seems more intricate that I thought it would be. I'm a little confused about how to go about this now. 😅

Will read try to read up on how to achieve the suggested solution.

**Comment by soumikcv** [comment]
> We would need to support both root and global styles dependent on the key

Did you mean the key in component object?

**Comment by siddharthkp** [comment]
oh nope, my bad. I meant the css selector - like body vs button

**Comment by soumikcv** [comment]
Ohno, i'm still confused. 
Do you want the API to look something like this right?

```
const components = {
  /** Global */
  Global: {
    body: {
      background: 'grays.900'
    },
    ':root': {
      '--variable-color': 'hotpink'
    },

    // Style below to be attached to ThemeProvider Root Element
    color: 'text.body',
    a: {
      color: 'text.link'
    }
  }
  //...other components
}
```

In this case, the `body` and `:root` style goes to the Emotion global and the rest to be attached to the Theme provider root element right?

**Comment by siddharthkp** [comment]
Yeah that's the direction I'm thinking in. Not totally convinced if that's the right way to go. But I feel it could work

**Comment by soumikcv** [comment]
This API looks a little confusing to me. The end user might put the style meant for root div inside the `body` itself not knowing what the styles outside `body` are meant to do. On the first glance it's confusing which style's gonna be applied where.

Separate explicit `Global` and `Root` keys may be an option to educate the end-user better on how the theme works but i'm not sure about it either.
Another way could be merging styles inside body with the root div styles and apply it to every theme provider root except the first one but this one feels hacky as well.

Also, how can you tell if a given `ThemeProvider` is occurring for the first time in the Component Tree implicitly? 
That information is needed because subsequent `ThemeProvider`s should not be injecting the body styles to global.

**Comment by soumikcv** [comment]
I was think about this earlier, another school of thought here could be to get rid of the Global component token and instead provide the user with a Global component (from emotion core) which automatically interpolates with the theme.

This way we can give user back a little more control without having to think about this bug. (the theme provider can still have the global needed by reach)

(Also, I apologise if me constantly bombarding half baked ideas in this thread is kinda annoying haha 😅)

**Comment by siddharthkp** [comment]
That's not a bad idea, here's why head is at:

1. If we put it in theme, then you can add globals next to components. you don't need to use a different api
2. If you're just using one of the default themes and want to add a font, you might not want to use the theme api. so a global component would help
3. It's not always global, sometimes these styles should be scoped to a section. Should we add a scoped: bool to control this?
4. Would root/global styles always be just next to the themeprovider? if yes, does it make sense to pass it to the themeprovider as a prop?

**Comment by soumikcv** [comment]
imo, the core issue here is that tightly coupling global style, root styles and theme provider gives you a confusing API which could lead to bugs (kinda what is happening to the docs website).

My solution is to decouple global styles while still keeping theme-provider and root styles together

### Global Style
> * If we put it in theme, then you can add globals next to components. you don't need to use a different api
> * If you're just using one of the default themes and want to add a font, you might not want to use the theme api. so a global component would help

If we take the global style from component token, only the first theme-provider is supposed to use it or else it would be buggy.  So, internally we would have to keep track of which one is the first theme-provider. On top of that we have to solve the problem of identifying and separating global and root styles and also be worried about how to handle the root styles in the case it is not the first occurrence of theme-prov.

Alternatively, if we use a global component 
- the need to track theme-providers occurrence is removed
- the end user has to write the global styles somewhere at the end of the day. This kind of a global component has already been popularized by css-in-js libraries (and also fits the semantics better imo, component tokens should be for component)
- global style has been decoupled so now it'll be easier to think about root style separately(explained in next section)


### Root Style
> * It's not always global, sometimes these styles should be scoped to a section. Should we add a scoped: bool to control this?
> * Would root/global styles always be just next to the themeprovider? if yes, does it make sense to pass it to the themeprovider as a prop?

The root style however, can be coupled with theme-provider as they will always be together.  Combing your suggestions, we can add a style prop to themeProvider which will take the root styles if passed. Moreover, after decoupling from global style `scoped: bool` would not be required to track it. 


Here's what the API i'm thinking of would look like

```jsx
<ThemeProvider
  tokens={theme.tokens}
  components={theme.components}
>
  <Global
    styles={{
      body: {
        backgroundColor: 'App.backgroundColor',
        color: 'text.body',
        transition: 'background 500ms, color 500ms'
      }
    }}
  />


  {/* Section which needs it's own theme */}
  <ThemeProvider
    tokens={theme2.tokens}
    components={theme2.components}
    style={{ a: { color: 'text.subtle' } }}
  >
    {/* Will have a div element with root styles on them */}
  </ThemeProvider>

  <ThemeProvider
    tokens={theme3.tokens}
    components={theme3.components}
  >
    {/* No need for root styles, hence no div element */}
  </ThemeProvider>
</ThemeProvider>
```
